### PR TITLE
Card comp

### PR DESCRIPTION
**Pull request by psergey**

### CODE REVIEW

File: src/components/Card.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { FC } from 'react';
+
+import classes from './Card.module.css';
+import { CardProps } from './Card.types';
+
+const MAX_TEXT_LENGTH = 80;
```
Comment (YarynaN) [inline src/components/Card.tsx]: NIT: I prefer to have separate file for constants. Just a small separation of concerns.

File: src/components/Card.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  tags,
+  isDragging,
+  maxTextSize = MAX_TEXT_LENGTH,
+}): React.ReactElement => {
+  const cardTags: string[] = tags?.split(',').slice(0, 2) ?? [];
+
+  return (
+    <div className={`${classes.card} ${isDragging ? classes.dragging : ''}`}>
+      <div className={classes['card-media']}>
+        <img src={image} alt={title}></img>
```
Comment (YarynaN) [inline src/components/Card.tsx]: Love that you did not forget alt! 🤌🏻
If you have html tag without children, you can for sure use self-closing ``

File: src/components/Card.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      </div>
+      <div className={classes['card-details']}>
+        <h2 className={classes['card-title']}>{title}</h2>
+        <p className={classes['card-description']}>
+          {text.length > maxTextSize ? text.substring(0, maxTextSize).trimEnd() + '...' : text}
```
Comment (YarynaN) [inline src/components/Card.tsx]: Okay that's one way of doing that. But I'd suggest https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow
The rule of thumb is - whatever you can do without js by means of css and html - do it with css and html.

File: src/components/CardList.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import { CardListItemProps, CardListProps } from './CardList.types';
+import Modal from './Modal';
+import ItemDetails from './ItemDetails';
+import '../utils/array';
+
+const CardList: FC<CardListProps> = ({ cards, onUpdated }): React.ReactElement => {
+  const [dragginItemId, setDragginItemId] = useState('');
+  const [dragOver, setDragOver] = useState('');
+
+  const handleDragStart = (e: React.DragEvent<HTMLElement>): void => {
```
Comment (YarynaN) [inline src/components/CardList.tsx:13]: Here we go! Please read https://react.dev/reference/react/useCallback I believe that's one of a good use cases (all of the callbacks here).

File: src/components/CardList.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    setDragOver(e.currentTarget.id);
+  };
+
+  const handleDragEnd = (): void => {
+    setDragginItemId('');
+    setDragOver('');
+  };
+
+  const handleDrop = (e: React.DragEvent<HTMLElement>): void => {
+    const srcIndex = cards.findIndex((item) => item.id === dragginItemId);
```
Comment (YarynaN) [inline src/components/CardList.tsx]: I'd suggest naming like `startIndex` and `endIndex` but it's NIT

File: src/components/CardList.tsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            <Card
+              key={item.id}
+              title={item.title}
+              image={item.image}
+              text={item.text}
+              tags={item.tags}
+              isDragging={item.id === dragginItemId}
+            />
+
+            {/* <Modal>
```
Comment (YarynaN) [inline src/components/CardList.tsx]: Please do not include commented code in PRs. :)

File: src/components/Modal.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  openName: '',
+  open: function (): void {},
+  close: function (): void {},
+});
+
+const Modal: FC<PropsWithChildren> = ({ children }): React.ReactElement => {
+  const [openName, setOpenName] = useState('');
+  const close = () => {
+    setOpenName('');
+    //document.body.style.overflow = '';
```
Comment (YarynaN) [inline src/components/Modal.tsx:23]: no commented code please.

File: src/components/Modal.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+import { useOutsideClick } from '../hooks/useOutsideClick';
+
+interface ModalContextType {
+  openName: string;
+  open: (name: string) => void;
+  close: () => void;
+}
+
+const ModalContext = createContext<ModalContextType>({
+  openName: '',
```
Comment (YarynaN) [inline src/components/Modal.tsx:14]: I would go for passing the whole object here not just a name.

File: src/components/Modal.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
+import { FC, PropsWithChildren, createContext, useContext, useState } from 'react';
+import { createPortal } from 'react-dom';
+
+import classes from './Modal.module.css';
+import { useOutsideClick } from '../hooks/useOutsideClick';
+
+interface ModalContextType {
```
Comment (YarynaN) [inline src/components/Modal.tsx:7]: The overall solution looks overcomplicated. This either needs some better code splitting or let's re-evaluate solution itself.

File: src/index.css [hunk 10]
```diff
@@ -1,23 +1,60 @@
...
+}
+
 :root {
   font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
   line-height: 1.5;
   font-weight: 400;
   font-size: 62.5%;
 
+  --color-grey-0: #fff;
+  --backdrop-color: rgba(255, 255, 255, 0.1);
```
Comment (YarynaN) [inline src/index.css:14]: Love global variables!

File: src/stories/CardList.stories.tsx
Comment (YarynaN) [inline src/stories/CardList.stories.tsx:1]: I would go for also a separate story for card itself 👍🏻

File: src/utils/array.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+interface Array<T> {
+  swap(srcIndex: number, destIndex: number): Array<T>;
+}
+
+Array.prototype.swap = function swap<T>(this: T[], srcIndex: number, destIndex: number): Array<T> {
```
Comment (YarynaN) [inline src/utils/array.ts]: This should work, but is there any particular need of creating polyfill and not just regular helper function?

**Review by YarynaN — COMMENTED** [review]
Overall this is good work! 💫
I left couple of comments that should help you flesh out your work. 
In next PR let's focus on the modal. I would suggest delete what's already there and start over fresh, but this is up to you of course.
Please, include Card story here.

File: src/utils/array.ts (hunk 11 above)
Comment (psergey) [reply src/utils/array.ts]: changed to regular function

File: src/components/CardList.tsx (hunk 5 above)
Comment (psergey) [reply src/components/CardList.tsx]: variable names were update

File: src/components/CardList.tsx (hunk 6 above)
Comment (psergey) [reply src/components/CardList.tsx]: cleanup

File: src/components/Card.tsx (hunk 2 above)
Comment (psergey) [reply src/components/Card.tsx]: fixed

**Pull request merged**